modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/editor/BalloonEditor.js

```diff
@@ -13,96 +13,176 @@
  */
 
 import PropTypes from 'prop-types';
-import React from 'react';
+import React, {useCallback, useEffect, useRef} from 'react';
 
 import {Editor} from './Editor';
 import DEFAULT_BALLOON_EDITOR_CONFIG from './config/DefaultBalloonEditorConfiguration';
 
 import '../css/main.scss';
 
-const BalloonEditor = ({config = {}, contents, name, ...otherProps}) => {
-	const editorConfig = {
-		...DEFAULT_BALLOON_EDITOR_CONFIG,
-		...config,
-	};
-
-	return (
-		<Editor
-			config={editorConfig}
-			data={contents}
-			name={name}
-			onBeforeLoad={(CKEDITOR) => {
-				CKEDITOR.ADDITIONAL_RESOURCE_PARAMS = {
-					languageId: themeDisplay.getLanguageId(),
-				};
-
-				CKEDITOR.disableAutoInline = true;
-
-				CKEDITOR.getNextZIndex = function () {
-					return CKEDITOR.dialog._.currentZIndex
-						? CKEDITOR.dialog._.currentZIndex + 10
-						: Liferay.zIndex.WINDOW + 10;
-				};
-			}}
-			onInstanceReady={(event) => {
-				const editor = event.editor;
-
-				const editable = editor.editable();
+const EMPTY_OBJECT = {};
+
+const BalloonEditor = React.forwardRef(
+	(
+		{
+			config = EMPTY_OBJECT,
+			contents,
+			name,
+			onChange,
+			onChangeMethodName,
+			...otherProps
+		},
+		ref
+	) => {
+		const editorConfig = {
+			...DEFAULT_BALLOON_EDITOR_CONFIG,
+			...config,
+		};
+
+		const editorRef = useRef();
+
+		const getHTML = useCallback(() => {
+			let data = contents;
+
+			const editor = editorRef.current.editor;
+
+			if (editor && editor.instanceReady) {
+				data = editor.getData();
+
+				if (
+					CKEDITOR.env.gecko &&
+					CKEDITOR.tools.trim(data) === '<br />'
+				) {
+					data = '';
+				}
 
-				// `floatPanel` plugin requires `id` to be `cke_${editor.name}`
+				data = data.replace(/(\u200B){7}/, '');
+			}
 
-				editable.setAttribute('id', `cke_${editor.name}`);
+			return data;
+		}, [contents]);
```

Comment:
This whole block is duplicated from `ClassicEditor`. Would it make sense to place it in `Editor`, as both editors are wrappers around it?

Alternatively, we could make a utility that both editors invoke.

Reply:
> Alternatively, we could make a utility that both editors invoke.

I was thinking to separate in utils like:

```js
import {getHTML} from 'frontend-editor-ckeditor-web';
```

and then reuse it. However, It is strictly internal for us and I'm afraid of ppl trying to import those functions and using externally

Reply:
You can import internally by using a relative path, like [here](https://github.com/liferay/liferay-portal/blob/master/modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/editor/BalloonEditor.js#L18).

Reply:
> Would it make sense to place it in Editor, as both editors are wrappers around it? 

Now I understood, you mean placing this useEffect inside Editor component and when using it, all editors will place a "common" window instance for calling functions and getting values

Reply:
https://github.com/liferay-frontend/liferay-portal/pull/1663/commits/8c86ae81a123ffc87b8b0ac25631b3dc2093ff1b